### PR TITLE
PMM-7296: ListAlertRules pagination

### PR DESCRIPTION
**Pull request by oter**
[PMM-7296](https://jira.percona.com/browse/PMM-7296)

- [x] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1457

- [x] API PR: https://github.com/percona/pmm/pull/662
- [x] Tests PR: https://github.com/Percona-Lab/pmm-api-tests/pull/180
---
- [x] Tests passed.
- [x] Feature build pass.
- [x] (Re)requested review.
- [x] Fix conflicts with target branch.
- [ ] Update API dependency.

### CODE REVIEW

File: models/rule_helpers.go [hunk 1]
```diff
@@ -58,6 +59,32 @@ func FindRules(q *reform.Querier) ([]*Rule, error) {
 	return rules, nil
 }
 
+// FindRulesOnPage returns a page with saved alert rules configuration.
+func FindRulesOnPage(q *reform.Querier, pageIndex, pageSize int) ([]*Rule, error) {
+	tail := fmt.Sprintf("ORDER BY id LIMIT %d OFFSET %d", pageSize, pageIndex*pageSize)
```
Comment (oter) [inline models/rule_helpers.go:64]: Probably need to change ordering -> by name and add an index for it.

File: services/management/ia/rules_service.go [hunk 2]
```diff
@@ -255,11 +256,88 @@ func (s *RulesService) ListAlertRules(ctx context.Context, req *iav1beta1.ListAl
...
+	}
+
+	res, err := s.convertAlertRules(rules, channels)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	totals := &iav1beta1.PageTotals{
+		TotalItems: int32(totalItems),
+		TotalPages: int32(totalItems) / page.Size,
```
Comment (Dasio) [inline services/management/ia/rules_service.go:337]: Truncation could happen. You need to round up.
101(items)/20(page size) => should be 6, not 5

File: models/rule_helpers.go [hunk 3]
```diff
@@ -58,6 +59,32 @@ func FindRules(q *reform.Querier) ([]*Rule, error) {
...
+	rows, err := q.SelectAllFrom(RuleTable, tail)
```
Comment (Dasio) [inline models/rule_helpers.go:65]:
```suggestion
	rows, err := q.SelectAllFrom(RuleTable, "ORDER BY id LIMIT $1 OFFSET $2", pageSize, pageIndex*pageSize)
```

**Review by Dasio — CHANGES_REQUESTED** [review]

File: services/management/ia/rules_service.go [hunk 4]
```diff
@@ -255,11 +256,88 @@ func (s *RulesService) ListAlertRules(ctx context.Context, req *iav1beta1.ListAl
 		return nil, status.Errorf(codes.FailedPrecondition, "%v.", services.ErrAlertingDisabled)
 	}
 
-	res, err := s.getAlertRules()
+	if req.Page == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Page should be set")
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/ia/rules_service.go:260]: I see two options here - either return all the items or set default page size (e.g. 20 items) and return first page. What do you think?

[the remark this answers is not in the document]
Comment (Dasio) [reply services/management/ia/rules_service.go:260]: Default page size seems better option for me.
But still, it is breaking change in API. But for now during beta it shouldn't be issue.

**Comment by Dasio** [comment]
Sorry, I my code suggestion is 'wrong', you need delete now tail variable because it is unused.

File: services/management/ia/rules_service.go (hunk 2 above)
Comment (oter) [reply services/management/ia/rules_service.go]: done, thanks.

File: services/management/ia/rules_service.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (oter) [reply services/management/ia/rules_service.go]: reworked

File: utils/utils.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package utils
+
+import "sort"
+
+func SetToSlice(set map[string]struct{}) []string {
```
[the remark this answers is not in the document]
Comment (oter) [reply utils/utils.go]: comment added

**Review by Dasio — APPROVED** [review]

File: services/management/ia/rules_service.go [hunk 6]
```diff
@@ -255,11 +256,100 @@ func (s *RulesService) ListAlertRules(ctx context.Context, req *iav1beta1.ListAl
...
-	res, err := s.getAlertRules()
+	const defaultPageSize = 20
+
+	var pageIndex, pageSize int
+	if req.Page == nil {
+		pageSize = defaultPageSize
+	} else {
+		pageIndex = int(req.Page.Index)
+		pageSize = int(req.Page.Size)
+	}
```
Comment (artemgavrilov) [inline services/management/ia/rules_service.go:267]: IMHO without `else` it's easier to read.

```suggestion
	pageIndex = 0 // first page
	pageSize =  defaultPageSize
	if req.Page != nil {
		pageIndex = int(req.Page.Index)
		pageSize = int(req.Page.Size)
	}
```

Comment (oter) [reply services/management/ia/rules_service.go]: for sure

**Review by artemgavrilov — APPROVED** [review]

File: models/rule_helpers.go (hunk 1 above)
Comment (oter) [reply models/rule_helpers.go]: out of scope for now

File: services/management/ia/rules_service.go [hunk 7]
```diff
@@ -250,11 +251,99 @@ func (s *RulesService) ListAlertRules(ctx context.Context, req *iav1beta1.ListAl
 		return nil, status.Errorf(codes.FailedPrecondition, "%v.", services.ErrAlertingDisabled)
 	}
 
-	res, err := s.getAlertRules()
+	const defaultPageSize = 20
```
Comment (BupycHuk) [inline services/management/ia/rules_service.go:254]: let's move it to the top of the file

Comment (oter) [reply services/management/ia/rules_service.go]: done

**Comment by BupycHuk** [comment]
@oter CI tests are failing

**Comment by oter** [comment]
@BupycHuk could you please take a look one more time, all the tests and FB are passing.

**Review by BupycHuk — APPROVED** [review]

**Comment by AlekSi** [comment]
@oter please resolve merge conflicts

**Pull request merged**